endpoint_route_handler/registry.py

```diff
@@ -0,0 +1,342 @@
+from odoo import tools
+from odoo.tools import DotDict
+
+from odoo.addons.base.models.ir_model import query_insert
+
+from .exceptions import EndpointHandlerNotFound
+
+
+def query_multi_update(cr, table_name, rows, cols):
+    """Update multiple rows at once.
+
+    :param `cr`: active db cursor
+    :param `table_name`: sql table to update
+    :param `rows`: list of dictionaries with write-ready values
+    :param `cols`: list of keys representing columns' names
+    """
+    # eg: key=c.key, route=c.route
+    keys = sql.SQL(",").join([sql.SQL("{0}=c.{0}".format(col)) for col in cols])
+    col_names = sql.SQL(",").join([sql.Identifier(col) for col in cols])
+    template = (
+        sql.SQL("(")
+        + sql.SQL(",").join([sql.SQL("%({})s".format(col)) for col in cols])
+        + sql.SQL(")")
+    )
+    query = sql.SQL(
+        """
+    UPDATE {table} AS t SET
+        {keys}
+    FROM (VALUES {values})
+        AS c({col_names})
+    WHERE c.key = t.key
+    RETURNING t.key
+    """
+    ).format(
+        table=sql.Identifier(table_name),
+        keys=keys,
+        col_names=col_names,
+        values=sql.Placeholder(),
+    )
+    execute_values(
+        cr,
+        query.as_string(cr._cnx),
+        rows,
+        template=template.as_string(cr._cnx),
+    )
+
+
+class EndpointRegistry:
+    """Registry for endpoints.
+
+    Used to:
+
+    * track registered endpoints
+    * retrieve routing rules to load in ir.http routing map
+    """
+
+    __slots__ = "cr"
+    _table = "endpoint_route"
+    _columns = (
+        # name, type, comment
+        ("key", "VARCHAR", ""),
+        ("route", "VARCHAR", ""),
+        ("opts", "text", ""),
+        ("routing", "text", ""),
+        ("endpoint_hash", "VARCHAR(32)", ""),
+        ("route_group", "VARCHAR(32)", ""),
+        ("updated_at", "TIMESTAMP NOT NULL DEFAULT NOW()", ""),
+    )
+
+    @classmethod
+    def registry_for(cls, cr):
+        return cls(cr)
+
+    @classmethod
+    def wipe_registry_for(cls, cr):
+        cr.execute("TRUNCATE endpoint_route")
+
+    @classmethod
+    def _setup_table(cls, cr):
+        if not tools.sql.table_exists(cr, cls._table):
+            tools.sql.create_model_table(cr, cls._table, columns=cls._columns)
+            tools.sql.create_unique_index(
+                cr,
+                "endpoint_route__key_uniq",
+                cls._table,
+                [
+                    "key",
+                ],
+            )
+            tools.sql.add_constraint(
+                cr,
+                cls._table,
+                "endpoint_route__endpoint_hash_uniq",
+                "unique(endpoint_hash)",
+            )
+
+            cr.execute(
+                """
+                CREATE OR REPLACE FUNCTION endpoint_route_set_timestamp()
+                    RETURNS TRIGGER AS $$
+                BEGIN
+                    NEW.updated_at = NOW();
+                    RETURN NEW;
+                END;
+                $$ LANGUAGE plpgsql;
```

Comment:
Here we should create a sequence `endpoint_version` and add a trigger to auto increment the sequence on update, insert and delete.